admin/autoupdater/src/autoupdater.c

```diff
@@ -0,0 +1,411 @@
+		 the latter is the case and update anyways as we can't do anything better
+		*/
+		fputs("autoupdater: warning: clock seems to be incorrect.\n", stderr);
+
+		if (get_uptime() < 600)
+			/* If the uptime is very low, it's possible we just didn't get the time over NTP yet, so we'll just wait until the next time the updater runs */
+			return 0;
+		else
+			/*
+			 Will give 1 when priority == 0, and lower probabilities the higher
+			 the priority value is (similar to the old static probability system)
+			*/
+			return powf(0.75f, priority);
+	}
+	else if (G.settings.fallback) {
+		if (diff >= seconds + 86400)
+			return 1;
+		else
+			return 0;
+	}
+	else if (diff >= seconds) {
+		return 1;
+	}
+	else {
+		float x = diff/seconds;
+
+		/*
+		 This is the simplest polynomial with value 0 at 0, 1 at 1, and which has a
+		 first derivative of 0 at both 0 and 1 (we all love continuously differentiable
+		 functions, right?)
+		*/
+		return 3*x*x - 2*x*x*x;
+	}
+}
+
+
+/** Receives data from uclient, chops it to lines and hands it to \ref parse_line */
+static void recv_manifest_cb(struct uclient *cl) {
+	static char buf[MAX_LINE_LENGTH + 1];
+	static char *ptr = buf;
+	char *newline;
+	int len;
+
+	while (1) {
+		if (ptr - buf == MAX_LINE_LENGTH) {
+			fputs("autoupdater: error: encountered manifest line exceeding limit of " STRINGIFY(MAX_LINE_LENGTH) " characters\n", stderr);
+			break;
+		}
+		len = uclient_read(cl, ptr, MAX_LINE_LENGTH - (ptr - buf));
+		if (!len)
+			break;
+		ptr[len] = '\0';
+
+		char *line = buf;
+		while (1) {
+			newline = strchr(line, '\n');
+			if (newline == NULL)
+				break;
+			*newline = '\0';
+
+			parse_line(line, &G.manifest, G.settings.branch);
+			line = newline + 1;
+		}
+
+		// Move the beginning of the next line to the beginning of the
+		// buffer. We cannot use strcpy here because the memory areas
+		// might overlap!
+		int n = strlen(line);
+		memmove(buf, line, n);
+		ptr = buf + n;
+	}
+}
+
+
+/** Receives data from uclient and writes it to file */
+static void recv_image_cb(struct uclient *cl) {
+	char buf[1024];
+	int len;
+
+	while (1) {
+		len = uclient_read(cl, buf, sizeof(buf));
+		if (!len)
+			return;
+
+		if (write(G.firmware_fd, buf, len) < len) {
+			fputs("autoupdater: error: downloading firmware image failed: ", stderr);
+			perror(NULL);
+			return;
+		}
+		ecdsa_sha256_update(&G.hash_ctx, buf, len);
+	}
+}
+
+
+static bool autoupdate(const char *mirror) {
+	/**** Get and check manifest *****************************************/
+	/* Clean manifest data from earlier mirror */
+	free_manifest(&G.manifest);
+	memset(&G.manifest, 0, sizeof(G.manifest));
+
+	/* Construct manifest URL */
+	char manifest_url[strlen(mirror) + strlen(G.settings.branch) + 11];
+	sprintf(manifest_url, "%s/%s.manifest", mirror, G.settings.branch);
+
+	/* Download manifest */
+	ecdsa_sha256_init(&G.manifest.hash_ctx);
+	get_url(manifest_url, recv_manifest_cb);
```

Comment:
The download functions definitely need better error handling. Obviously the autoupdater will error out as soon as the checksums are checked, but "download failed" would be a much more useful error message.